scenario/runtime.py

```diff
@@ -298,38 +295,48 @@ def _virtual_charm_root(self) -> typing.ContextManager[Path]:
         config_yaml = virtual_charm_root / "config.yaml"
         actions_yaml = virtual_charm_root / "actions.yaml"
 
-        metadata_files_present = any(
-            file.exists() for file in (metadata_yaml, config_yaml, actions_yaml)
-        )
+        metadata_files_present: Dict[Path, Union[str, False]] = {
+            file: file.read_text() if file.exists() else False
+            for file in (metadata_yaml, config_yaml, actions_yaml)
+        }
+
+        any_metadata_files_present_in_vroot = any(metadata_files_present.values())
 
         if spec.is_autoloaded and vroot_is_custom:
             # since the spec is autoloaded, in theory the metadata contents won't differ, so we can
             # overwrite away even if the custom vroot is the real charm root (the local repo).
             # Still, log it for clarity.
-            if metadata_files_present:
-                logger.info(
+            if any_metadata_files_present_in_vroot:
+                logger.debug(
                     f"metadata files found in custom vroot {vroot}. "
                     f"The spec was autoloaded so the contents should be identical. "
                     f"Proceeding...",
                 )
 
-        elif not spec.is_autoloaded and metadata_files_present:
-            logger.error(
+        elif not spec.is_autoloaded and any_metadata_files_present_in_vroot:
+            logger.warn(
                 f"Some metadata files found in custom user-provided vroot {vroot} "
-                f"while you have passed meta, config or actions to trigger(). "
-                "We don't want to risk overwriting them mindlessly, so we abort. "
-                "You should not include any metadata files in the charm_root. "
-                "Single source of truth are the arguments passed to trigger(). ",
+                "while you have passed meta, config or actions to Context.run(). "
+                "Single source of truth are the arguments passed to Context.run(). "
+                "Vroot metadata files will be overwritten for the "
+                "duration of this test, and restored afterwards. "
+                "To avoid this, clean any metadata files from the vroot before calling run.",
             )
-            raise DirtyVirtualCharmRootError(vroot)
 
         metadata_yaml.write_text(yaml.safe_dump(spec.meta))
         config_yaml.write_text(yaml.safe_dump(spec.config or {}))
         actions_yaml.write_text(yaml.safe_dump(spec.actions or {}))
 
         yield virtual_charm_root
 
-        if not vroot_is_custom:
+        if vroot_is_custom:
+            for file, previous_content in metadata_files_present.items():
+                if not previous_content:  # False: file did not exist before
```

Comment:
Similar to above, probably better to be more explicit:

```suggestion
                if previous_content is None:  # None: file did not exist before
```